src/dune_engine/scheduler.ml

```diff
@@ -643,8 +676,101 @@ module Handler = struct
   type t = Config.t -> Event.t -> unit
 end
 
+module Timer : sig
+  type t
+
+  val create : Event.Queue.t -> frequency:float -> t
+
+  type sleeper
+
+  val await : sleeper -> [ `Finished | `Cancelled ] Fiber.t
+
+  val cancel : t -> sleeper -> unit Fiber.t
+
+  val sleep : t -> float -> sleeper
+
+  val close : t -> unit
+end = struct
+  type sleeper = [ `Finished | `Cancelled ] Fiber.Ivar.t
+
+  type t =
+    { events : Event.Queue.t
+    ; mutex : Mutex.t
+    ; frequency : float
+    ; mutable sleepers : (float * [ `Finished | `Cancelled ] Fiber.Ivar.t) list
+    ; mutable active : bool
+    }
+
+  let await = Fiber.Ivar.read
+
+  let cancel t sleeper =
+    Mutex.lock t.mutex;
+    t.sleepers <-
+      List.filter t.sleepers ~f:(fun (_, sleeper') -> not (sleeper' == sleeper));
+    Mutex.unlock t.mutex;
+    let* res = Fiber.Ivar.peek sleeper in
+    match res with
+    | None -> Fiber.Ivar.fill sleeper `Cancelled
```

Comment:
Should this happen if and only if the "filter" above removed at least one entry? Should we have a check? (or a comment)

Reply:
I suspect there may be a timer event in the scheduler queue at this time, which would mean that filling the ivar here will break the "fill" that happens in the scheduler.

Reply:
> I suspect there may be a timer event in the scheduler queue at this time, which would mean that filling the ivar here will break the "fill" that happens in the scheduler.

Yes indeed. I forgot about this race.